### ISSUE TITLE
Dummy question about concurrency...

### ISSUE THREAD

**Issue by jan-johansson-mr** [dotnet/orleans#3618]
Hi,

Assume we have an interface "IMyGrain", declared as follows:

        public partial interface IMyGrain : Orleans.IGrainWithIntegerKey

And that we have "MyGrain" (implementing the interface) with its context (string, or whatever).

Later when we need to invoke "IMyGrain", we use:

        var friend = client.GetGrain<IMyGrain>(id);

Where the id argument is any integer, e.g. '0', to return for us a "proxy".

Now, if we have, let's say 10 "proxies", each with unique id 0, 1, 2, ..., 9 (different id for each grain), and then invoke each of these concurrently, do each grain execute concurrently, or is each invocation placed in queue for the grain to handle (regardless of the id)?

Conversely, if the 10 "proxies" were all related to the grain with the same id, e.g. 0, and then each "proxy" were invoked concurrently, is each invocation placed in queue for the grain to handle (it should, since in this case the same grain is referred through the same id) :-)

Thanks

**Comment by JanEggers** [comment]
> Now, if we have, let's say 10 "proxies"

you can configure maximum concurrency in your NodeConfiguration.

> if the 10 "proxies" were all related to the grain with the same id, e.g. 0

each stateful Grain can only exist once in the cluster. the orleans runtime manages this with so called ActivationData that knows if another task is currently executed for a grain and contains a list of messages not yet processed. 

so the runtime makes shure there is only one task active on your grain at a time. 
but if you need to handle concurrent changes of the same grain have a look at:
http://richorama.github.io/2015/01/22/optimistic-concurrency-in-orleans/

**Comment by jan-johansson-mr** [comment]
Thanks Jan (or is it JanEggers? - my name is Jan) :-)

Anyway, it's simple to verify.

I created two grains, one with id '0' and one with id '1', the grain with '0' has a long (simulated by sleep) execution time, while the grain with id '1' has almost zero execution time. When running, I see through the output (on screen) that while the grain with id '0' stalls, the other grain with id '1' flow through multiple calls.

By using id '0' on all grains, I learned that now all calls are stalled and executed in sequential manners, as I kind of expected, but verified.

Lesson from this is that grains are granular to context, arbitrated by the grain id, and this is amazing :-) and usable in so many ways.
